crates/common/tedge_utils/src/fs_notify.rs

```diff
@@ -0,0 +1,600 @@
+        let mut set = btreeset! {};
+        let hash_map = self.description.get(&dir_path)?;
+
+        for masks in hash_map.values() {
+            for mask in masks {
+                set.insert(mask.to_owned());
+            }
+        }
+
+        let v = Vec::from_iter(set);
+        Some(v)
+    }
+}
+
+pub struct NotifyStream {
+    buffer: [u8; 1024],
+    inotify: Inotify,
+    watchers: WatchDescriptor,
+}
+
+impl TryDefault for NotifyStream {
+    type Error = NotifyStreamError;
+
+    fn try_default() -> Result<Self, Self::Error> {
+        let inotify = Inotify::init();
+        match inotify {
+            Ok(inotify) => {
+                let buffer = [0; 1024];
+                Ok(Self {
+                    buffer,
+                    inotify,
+                    watchers: WatchDescriptor::default(),
+                })
+            }
+            Err(err) => Err(NotifyStreamError::FromIOError(err)),
+        }
+    }
+}
+
+/// normalisation step joining `candidate_watch_dir` and `candidate_file` and computing the parent of `candidate_file`.
+///
+/// this is useful in situations where:
+/// `candidate_watch_dir` = /path/to/a/directory
+/// `candidate_file` = continued/path/to/a/file
+///
+/// this function will concatenate the two, into:
+/// `/path/to/a/directory/continued/path/to/a/file`
+/// and will return:
+/// `/path/to/a/directory/continued/path/to/a/` and `file`
+fn normalising_watch_dir_and_file(
+    candidate_watch_dir: &Path,
+    candidate_file: Watch,
+) -> Result<(PathBuf, Watch), NotifyStreamError> {
+    match candidate_file {
+        Watch::File(file_name) => {
+            let full_path = candidate_watch_dir.join(file_name);
+            let full_path = &full_path;
+            let parent =
+                full_path
+                    .parent()
+                    .ok_or_else(|| NotifyStreamError::ErrorNormalisingWatcher {
+                        path: full_path.to_path_buf(),
+                    })?;
+            let file = full_path
+                .file_name()
+                .and_then(|f| f.to_str())
+                .ok_or_else(|| NotifyStreamError::ErrorNormalisingWatcher {
+                    path: full_path.to_path_buf(),
+                })?;
+            Ok((parent.to_path_buf(), Watch::File(file.to_string())))
+        }
+        Watch::Directory => Ok((candidate_watch_dir.to_path_buf(), Watch::Directory)),
+    }
+}
+
+/// to allow notify to watch for multiple events (CLOSE_WRITE, CREATE, MODIFY, etc...)
+/// our internal enum `Masks` needs to be converted into a single `WatchMask` via bitwise OR
+/// operations. (Note, our `Masks` type is an enum, `WatchMask` is a bitflag)
+pub(crate) fn pipe_masks_into_watch_mask(
+    masks: &[FileEvent],
+) -> Result<WatchMask, NotifyStreamError> {
+    let mut watch_mask = WatchMask::empty();
+    for mask in masks {
+        watch_mask |= mask.clone().try_into()?
```

Comment:
sadly removing the clone would need a to_owned because `mask` is a referenced there...

the TryFrom plays the role to limit what the user of the API can "watch". See the above explanation [here](https://github.com/thin-edge/thin-edge.io/pull/1244#discussion_r921450910)

Reply:
Ah,... yes. :disappointed: I missed that part. Sorry!